providers/session_state_test.go

```diff
@@ -153,3 +133,182 @@ func TestExpired(t *testing.T) {
 	s = &SessionState{}
 	assert.Equal(t, false, s.IsExpired())
 }
+
+type testCase struct {
+	SessionState
+	Encoded string
+	Cipher  *cookie.Cipher
+	Error   bool
+}
+
+// TestEncodeSessionState tests EncodeSessionState with the test vector
+//
+// - Currently only tests without cipher here because we have no way to mock
+// the random generator used in EncodeSessionState.
+// - The zero value of time.Time is encoded to "0001-01-01T00:00:00Z"
```

Comment:
Agreed.  One easy workaround would be declaring `ExpiresOn` as `*time.Time`.  Extra nil checks will be needed in `IsExpired()` and other places where it gets referenced.

Reply:
Yeah that should work, do you have time to try it?

Reply:
Pushed my first attempt.  Introduced timeVal/timePtr helper functions and it looks a bit cumbersome.  How do you feel on it?

Reply:
Ok, here's another approach.  It utilizes struct embedding.  I think this is preferable because there's no cumbersome helper functions or side effects on provider's code.

The first attempt using timeVal/timePtr is preserved in https://github.com/yaegashi/oauth2_proxy/commits/session-state-json-2 for reference.